File: Combinations.py
class Solution:
	cache = dict()
	# @return a list of lists of integers
	def combine(self, n, k):
		# return [combine(n-1, k), [combine(n-1,k-1),n] ]
		if (not self.cache.has_key((n,k))):
			if (k == 0):
				self.cache[(n,k)] = [[]]
			elif (n == 0):
				self.cache[(n,k)] = [[]]
			elif (n == k):
				self.cache[(n,k)] = [range(1,n+1)]
			else:
				c1 = copy.deepcopy(self.combine(n-1, k-1))
				for i in xrange(len(c1)):
					c1[i].append(n)
				c1.extend(self.combine(n-1,k))
				self.cache[(n,k)] = c1
		return self.cache[(n,k)]

File: Combinations.java
public class Solution {
    Map<Integer, Map<Integer, List<List<Integer>>>> cache = new HashMap<Integer, Map<Integer, List<List<Integer>>>>();
    public List<List<Integer>> combine(int n, int k) {
        // C(n, k) = C(n-1, k-1) + C(n-1, k)
        if (!cache.containsKey(n)) {
            cache.put(n, new HashMap<Integer, List<List<Integer>>>());
        }
        if (!cache.containsKey(n-1)) {
            cache.put(n-1, new HashMap<Integer, List<List<Integer>>>());
        }
        if (!cache.get(n).containsKey(k)) {
            List<List<Integer>> c = new ArrayList<List<Integer>>();
            if (k == 0 || n == 0) {
                List<Integer> e = new ArrayList<Integer>();
                c.add(e);
            }
            else if (n == k) {
                List<Integer> e = new ArrayList<Integer>();
                for (int i = 1; i <= n; i++)
                    e.add(i);
                c.add(e);
            }
            else {
                List<List<Integer>> c1 = combine(n-1, k-1);
                List<List<Integer>> c2 = combine(n-1, k);
                for (List<Integer> v : c1) {
                    List<Integer> vv = new ArrayList<Integer>(v);
                    vv.add(n);
                    c.add(vv);
                }
                for (List<Integer> v : c2) {
                    c.add(v);
                }
            }
            cache.get(n).put(k, c);
        }
        return cache.get(n).get(k);
    }
}